File: frontend-refactoring/src/services/ReservationService.js
import axios from 'axios';
import config from '@config';

const API_URL =  config.API_URL + '/reservations';

class ReservationService {
  async requestReservation(reservationRequestDTO) {
    try {
      const response = await axios.post(API_URL, reservationRequestDTO);
      return response.data;
    } catch (error) {
      console.error('Error creating reservation:', error);
      throw error;
    }
  }

  async getReservation(id) {
    try {
      const response = await axios.get(`${API_URL}/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching reservation with ID ${id}:`, error);
      throw error;
    }
  }

  async deleteReservation(id) {
    try {
      await axios.delete(`${API_URL}/${id}`);
    } catch (error) {
      console.error(`Error deleting reservation with ID ${id}:`, error);
      throw error;
    }
  }

  async getAllReservations() {
    try {
      const response = await axios.get(API_URL);
      return response.data;
    } catch (error) {
      console.error('Error fetching all reservations:', error);
      throw error;
    }
  }

  async getReservationsByCafeIdAndDate(cafeId, date) {
    try {
      const isoDate = date.toISOString();
      console.log(cafeId, isoDate)
      const response = await axios.get(`${API_URL}/cafe-reservations`, {
        params: { cafeId, date: isoDate }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching reservations by cafe and date:', error);
      throw error;
    }
  }

}

export default new ReservationService();

File: backend/src/main/java/com/example/demo/service/ReservationService.java
package com.example.demo.service;

import com.example.demo.Exception.ResourceNotFoundException;
import com.example.demo.dto.ReservationDTO;
import com.example.demo.dto.ReservationRequestDTO;
import com.example.demo.entity.Cafe;
import com.example.demo.entity.Reservation;
import com.example.demo.entity.User;
import com.example.demo.mapper.ReservationMapper;
import com.example.demo.repository.CafeRepository;
import com.example.demo.repository.ReservationRepository;
import com.example.demo.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class ReservationService {

    @Autowired
    private ReservationRepository reservationRepository;

    @Autowired
    private CafeRepository cafeRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private ReservationMapper reservationMapper;

    @Transactional
    public ReservationDTO requestReservation(ReservationRequestDTO reservationRequestDTO) {
        // Validate cafeId and userId existence
        Cafe cafe = cafeRepository.findById(reservationRequestDTO.getCafeId())
                .orElseThrow(() -> new ResourceNotFoundException("Cafe not found with id: " + reservationRequestDTO.getCafeId()));

        User user = userRepository.findById(reservationRequestDTO.getUserId())
                .orElseThrow(() -> new ResourceNotFoundException("User not found with id: " + reservationRequestDTO.getUserId()));

        // Create Reservation entity
        Reservation reservation = Reservation.builder()
                .cafe(cafe)
                .user(user)
                .seatNumber(reservationRequestDTO.getSeatNumber())
                .startTime(reservationRequestDTO.getStartTime())
                .endTime(reservationRequestDTO.getEndTime())
                .build();

        // Save Reservation entity
        reservation = reservationRepository.save(reservation);

        // Map to DTO and return
        return reservationMapper.toDto(reservation);
    }

    @Transactional(readOnly = true)
    public ReservationDTO getReservation(Long id) {
        Reservation reservation = reservationRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Reservation not found with id: " + id));
        return reservationMapper.toDto(reservation);
    }

    @Transactional(readOnly = true)
    public List<ReservationDTO> getReservationsByCafeIdAndDate(Long cafeId, LocalDateTime date) {
        List<Reservation> reservations = reservationRepository.findReservationsByCafeIdAndDate(cafeId, date);
        return reservations.stream()
                .map(reservationMapper::toDto)
                .collect(Collectors.toList());
    }


    @Transactional
    public void deleteReservation(Long id) {
        Reservation reservation = reservationRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Reservation not found with id: " + id));
        reservationRepository.delete(reservation);
    }

    @Transactional(readOnly = true)
    public List<ReservationDTO> getAllReservations() {
        List<Reservation> review = reservationRepository.findAll();
        return review.stream()
                .map(reservationMapper::toDto)
                .collect(Collectors.toList());
    }
}
